### ISSUE TITLE
recovery_session vacuuming?

### ISSUE THREAD

**Issue by bodqhrohro** [luakit/luakit#775]
## I'm submitting a…
Other (mostly a question, but a combo of bug report and feature request too)

## Current Behavior
My `recovery_session` has already reached **more than 13 MB**, though it contains **only 200 tabs**. For comparison: a JSON session file **from SeaMonkey** profile that I used almost 2 years, that has almost **1000 tabs** and stores a lot of metadata besides of URLs, is **less than 6 MB**. LuaKit, OTOH, seems to use some ad-hoc format, and to store not the static dump but the **whole history of actions**. If so, can it be vacuumed, so it should contain only the **current state** of tabs?

I started having a performance issue some days before: the `luakit` process **usually** occupies near **200 MB** of RAM, but from time to time the memory usage **grows** to near **1 GB**, or even more, and then **back**. This causes **thrashing** if the system had no enough free memory for such an unexpected quake. I had no idea what is this caused by: thought that this may be some buffer sync issue with LuaKit and X.Org/Compiz, but inspecting shared memory buffers did not confirm that; tried to disable Gigacage, even though WebKit things don't seem to be related to `luakit` process at all. And finally, when I started typing an issue about it, LuaKit **crashed** with `unprotected error in call to Lua API (not enough memory)` error, and **corrupted the session file** (luckily, I made a backup right before). So now I think this happens when LuaKit writes the session file; if vacuuming won't help, I will post a separate issue.

## Expected Behavior
A compact session recovery file.

## Reproduction Instructions
Having a long-living tab-oholic session.

## Environment

Linux Distribution & Version: Debian 10
Output of `luakit --version`: luakit 2.0.0
  built with webkit 2.22.6 (installed version: 2.22.7)

**Comment by bodqhrohro** [comment]
The issue with RAM quakes seems to be a duplicate of https://github.com/luakit/luakit/issues/761

**Comment by bodqhrohro** [comment]
I investigated the file more deeply, and found out that my assumption about the complete action history is probably wrong. It's simply a mess.

First, in one of tabs I was filling a wiki, went through dozens of pages, and all of the text filled in was stored for months. That's pretty good that LuaKit takes care of POST data, so this is mostly my fault.

However, when I closed that tab and made sure that it went beyond the horizon of `undoclose.max_saved_tabs`, the session file became only a few megabytes smaller. There is still a lot of messy stuff: long URLs of ad frames, large pieces of HTML with some scripts. Looks like almost no one faced this issue before just because others do block ads/trackers :D Does LuaKit serialize the contents of all of frames? for what?

**Comment by bodqhrohro** [comment]
Here's an example of an URL that brought a lot of messy stuff into the corresponding `session_state` entry: https://www.fastcompany.com/90229453/get-ready-for-the-splinternet-the-web-might-not-be-worldwide-much-longer?utm_source=feedly&utm_medium=webfeeds

**Comment by bodqhrohro** [comment]
Top 10 of longest lines:
```
11227033
147116
86050
79729
55756
38261
33240
32433
21287
20687
```
11 MB line?!! Seriously?

Some strong AWK magic to determine what exactly:
```
bq:03:03:35:/data/office/tex$ awk '/^\["uri"\]/ { print sum; printf "%s ", $0; sum = 0 } { sum += length } END { print sum }' ~/.local/share/luakit/recovery_session 2>/dev/null |awk '{ print $2 " " $1 }'|sort -rn|head -n 1
11443448 ["uri"]="https://www.proz.com/forum/translation_in_ukraine_%D0%BF%D0%B5%D1%80%D0%B5%D0%BA%D0%BB%D0%B0%D0%B4_%D0%B2_%D1%83%D0%BA%D1%80%D0%B0%D1%97%D0%BD%D1%96/274528-%D1%80%D0%B5%D0%B3%D0%B8%D1%81%D1%82%D1%80%D0%B0%D1%86%D0%B8%D1%8F_%D1%84%D0%BB%D0%BF_%D0%B4%D0%BB%D1%8F_%D0%BF%D0%B5%D1%80%D0%B5%D0%B2%D0%BE%D0%B4%D1%87%D0%B8%D0%BA%D0%BE%D0%B2_%D0%B2_%D1%83%D0%BA%D1%80%D0%B0%D0%B8%D0%BD%D0%B5_%D0%B2%D0%B0%D1%88%D0%B5_%D0%BC%D0%BD%D0%B5%D0%BD%D0%B8%D0%B5-page3.html",
```
https://www.proz.com/forum/translation_in_ukraine_%D0%BF%D0%B5%D1%80%D0%B5%D0%BA%D0%BB%D0%B0%D0%B4_%D0%B2_%D1%83%D0%BA%D1%80%D0%B0%D1%97%D0%BD%D1%96/274528-%D1%80%D0%B5%D0%B3%D0%B8%D1%81%D1%82%D1%80%D0%B0%D1%86%D0%B8%D1%8F_%D1%84%D0%BB%D0%BF_%D0%B4%D0%BB%D1%8F_%D0%BF%D0%B5%D1%80%D0%B5%D0%B2%D0%BE%D0%B4%D1%87%D0%B8%D0%BA%D0%BE%D0%B2_%D0%B2_%D1%83%D0%BA%D1%80%D0%B0%D0%B8%D0%BD%D0%B5_%D0%B2%D0%B0%D1%88%D0%B5_%D0%BC%D0%BD%D0%B5%D0%BD%D0%B8%D0%B5-page3.html

**Comment by aidanholm** [comment]
Yeah, that's a little excessive. Luakit does serialize the contents of all tabs, mostly to support the undoclose feature, but also to save other state like the current scroll position for all tabs. The undoclose feature then stores all of its state directly in the session file, even though serialized tabs probably lose some of their benefits between sessions.

I've never seen serialized tabs that large, even with my adblocking disabled, js enabled, etc. However, I think it would be reasonable to strip the serialized tab data from saved sessions.
